### PR TITLE
Chat/initial

### PR DESCRIPTION
**Pull request by NikAshanin**

### CODE REVIEW

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 1]
```diff
@@ -0,0 +1,253 @@
+//
+//  KeyboardTracker.swift
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: put it in `LeadKit/LeadKit/Classes/Chat/`

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+//  LeadKit
+//
+//  Created by Николай Ашанин on 15.09.16.
+//  Copyright © 2016 Touch Instinct. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+class KeyboardTracker {
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: public

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 3]
```diff
@@ -0,0 +1,253 @@
+//
+//  KeyboardTracker.swift
+//  LeadKit
+//
+//  Created by Николай Ашанин on 15.09.16.
+//  Copyright © 2016 Touch Instinct. All rights reserved.
+//
+
+import Foundation
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: can be removed

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    private enum KeyboardStatus {
+        case Hidden
+        case Showing
+        case Shown
+    }
+
+    private var keyboardStatus: KeyboardStatus = .Hidden
+    private let view: UIView
+    private let inputContainerBottomConstraint: NSLayoutConstraint
+    var trackingView: UIView {
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: public?

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    }
+    private lazy var keyboardTrackerView: KeyboardTrackingView = {
+        let trackingView = KeyboardTrackingView()
+        trackingView.positionChangedCallback = { [weak self] in
+            self?.layoutInputAtTrackingViewIfNeeded()
+        }
+        return trackingView
+    }()
+
+    var isTracking = false
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: public?

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    var inputContainer: UIView
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: public?

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    private var notificationCenter: NSNotificationCenter
+
+    init(viewController: UIViewController, inputContainer: UIView,
+         inputContainerBottomContraint: NSLayoutConstraint, notificationCenter: NSNotificationCenter) {
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]:
```swift
notificationCenter: NSNotificationCenter = NSNotificationCenter.defaultCenter()
```

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+            self?.layoutInputAtTrackingViewIfNeeded()
+        }
+        return trackingView
+    }()
+
+    var isTracking = false
+    var inputContainer: UIView
+    private var notificationCenter: NSNotificationCenter
+
+    init(viewController: UIViewController, inputContainer: UIView,
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: `viewController` itself is not required pass `viewControllerView`

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+         inputContainerBottomContraint: NSLayoutConstraint, notificationCenter: NSNotificationCenter) {
+        view = viewController.view
+        inputContainer = inputContainer
+        inputContainerBottomConstraint = inputContainerBottomContraint
+        self.notificationCenter = notificationCenter
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: self is not required

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+        self.notificationCenter.addObserver(self,
+                                            selector: #selector(KeyboardTracker.keyboardWillShow(_:)),
+                                            name: UIKeyboardWillShowNotification, object: nil)
+        self.notificationCenter.addObserver(self,
+                                            selector: #selector(KeyboardTracker.keyboardDidShow(_:)),
+                                            name: UIKeyboardDidShowNotification, object: nil)
+        self.notificationCenter.addObserver(self,
+                                            selector: #selector(KeyboardTracker.keyboardWillHide(_:)),
+                                            name: UIKeyboardWillHideNotification, object: nil)
+        self.notificationCenter.addObserver(self,
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: the same about all self

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    private func keyboardWillHide(notification: NSNotification) {
+        guard isTracking else {
+            return
+        }
+        keyboardStatus = .Hidden
+        layoutInputAtBottom()
+    }
+
+    private func bottomConstraintFromNotification(notification: NSNotification) -> CGFloat {
+        guard let rect = (notification.userInfo?[UIKeyboardFrameEndUserInfoKey] as? NSValue)?.CGRectValue() else {
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: add extension for NSNotification with computed optional properties

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        }
+        adjustTrackingViewSizeIfNeeded()
+    }
+
+    private func adjustTrackingViewSizeIfNeeded() {
+        let inputContainerHeight = inputContainer.bounds.height
+        let trackerViewHeight = trackingView.bounds.height
+        if trackerViewHeight != inputContainerHeight {
+            keyboardTrackerView.bounds = CGRect(origin: CGPoint.zero,
+                                                size: CGSize(width: self.keyboardTrackerView.bounds.width,
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: self is not required

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+                                                        height: 0))
+        inputContainerBottomConstraint.constant = 0
+        view.layoutIfNeeded()
+    }
+
+    func layoutInputAtTrackingViewIfNeeded() {
+        guard isTracking && keyboardStatus == .Shown else {
+            return
+        }
+        let newBottomConstraint = bottomConstraintFromTrackingView()
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: `bottomConstraintFromTrackingView ` -> computed property

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+
+private class KeyboardTrackingView: UIView {
+
+    private static let centerKey = "center"
+
+    var positionChangedCallback: (() -> Void)?
+    var observedView: UIView?
+
+    deinit {
+        if let observedView = self.observedView {
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: self is not required

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+            }
+        }
+    }
+
+    private override func intrinsicContentSize() -> CGSize {
+        return bounds.size
+    }
+
+    override func willMoveToSuperview(newSuperview: UIView?) {
+        if let observedView = self.observedView {
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: self is not required

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+            return
+        }
+        if object === superview {
+            guard let sChange = change else {
+                return
+            }
+            let oldCenter = (sChange[NSKeyValueChangeOldKey] as? NSValue)?.CGPointValue()
+            let newCenter = (sChange[NSKeyValueChangeNewKey] as? NSValue)?.CGPointValue()
+            if oldCenter != newCenter {
+                self.positionChangedCallback?()
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: self is not required

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+                                                          context: UnsafeMutablePointer<Void>) {
+        guard let object = object, superview = self.superview else {
+            return
+        }
+        if object === superview {
+            guard let sChange = change else {
+                return
+            }
+            let oldCenter = (sChange[NSKeyValueChangeOldKey] as? NSValue)?.CGPointValue()
+            let newCenter = (sChange[NSKeyValueChangeNewKey] as? NSValue)?.CGPointValue()
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: add optional computed properties to `NSNotification`

File: LeadKit/LeadKit/Classes/KeyboardTracker.swift (hunk 8 above)
Comment (NikAshanin) [inline LeadKit/LeadKit/Classes/KeyboardTracker.swift]: KeyboardTracker is not viewcontroller. What you mean?

File: LeadKit/LeadKit/Classes/Chat/KeyboardTracker.swift [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+}
+
+private class KeyboardTrackingView: UIView {
+
+    private static let centerKey = "center"
+
+    var positionChangedCallback: (() -> Void)?
+    var observedView: UIView?
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/Chat/KeyboardTracker.swift]: public or private?

File: LeadKit/LeadKit/Classes/Chat/KeyboardTracker.swift [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+        view.layoutIfNeeded()
+    }
+
+}
+
+private class KeyboardTrackingView: UIView {
+
+    private static let centerKey = "center"
+
+    var positionChangedCallback: (() -> Void)?
```
Comment (petropavel13) [inline LeadKit/LeadKit/Classes/Chat/KeyboardTracker.swift]: public or private?

**Pull request closed without merging**